src/main/java/com/teragrep/cfe_39/consumers/kafka/DatabaseOutput.java

```diff
@@ -0,0 +1,396 @@
+    private final SDVector eventNodeSourceHostname;
+    private final SDVector eventNodeRelayHostname;
+    private final SDVector originHostname;
+    private File syslogFile;
+    private final Config config;
+    private long approximatedSize; // FIXME: Not working properly when flush() is not used after append in the AVRO-file.
+    private long epochMicros_last;
+
+    DatabaseOutput(
+            Config config,
+            String table,
+            DurationStatistics durationStatistics,
+            TopicCounter topicCounter
+    ) {
+        this.config = config;
+        this.table = table;
+        this.durationStatistics = durationStatistics;
+        this.topicCounter = topicCounter;
+        this.minimumFreeSpace = 32000000; // TODO: CHECK RIGHT VALUE FOR minimumFreeSpace
+        this.maximumFileSize = config.getMaximumFileSize();; // Maximum file size should be 64M (64000000). 60800000 is 95% of 64M which should be a good approximation point.
+
+        // queueDirectory and queueNamePrefix are only used for temporarily storing the AVRO-serialized files before committing them to HDFS when the file size reaches the threshold (or all records are processed).
+        this.writableQueue = new WritableQueue(
+                config.getQueueDirectory()
+        );
+
+        this.sourceConcatenationBuffer = ByteBuffer.allocateDirect(256*1024);
+        teragrepStreamName = new SDVector("teragrep@48577","streamname");
+        teragrepDirectory = new SDVector("teragrep@48577","directory");
+        this.eventNodeSourceSource = new SDVector("event_node_source@48577","source");
+        this.eventNodeRelaySource = new SDVector("event_node_relay@48577","source");
+        this.eventNodeSourceSourceModule = new SDVector("event_node_source@48577","source_module");
+        this.eventNodeRelaySourceModule = new SDVector("event_node_relay@48577","source_module");
+        this.eventNodeSourceHostname = new SDVector("event_node_source@48577","hostname");
+        this.eventNodeRelayHostname = new SDVector("event_node_relay@48577","hostname");
+        this.originHostname = new SDVector("origin@48577","hostname");
+        this.approximatedSize = 0;
+        this.epochMicros_last = 0L;
+    }
+
+    // Checks that the filesize stays under the defined maximum file size.
+    boolean checkSizeTooLarge(long fileSize, RecordOffsetObject recordOffsetObject) {
+        try {
+            // If the syslogAvroWriter is already initialized, check the filesize so it doesn't go above maximumFileSize.
+            if (fileSize > maximumFileSize) {
+                // file too large for adding the new record to it, write the still adequately sized AVRO-file to the HDFS database and create a new empty AVRO-file.
+
+                // This part closes the writing of now "complete" AVRO-file and stores the file to HDFS.
+                syslogAvroWriter.close();
+                try (HDFSWriter writer = new HDFSWriter(config, recordOffsetObject)) {
```

Comment:
Does this need to be closed manually?

Reply:
With the HDFSWriter writer I ran into an issue where Hadoop FileSystem.close() doesn't just affect the current class and thread, it affects all the FileSystem objects that were created using FileSystem.get(URI.create(hdfsuri), conf); in different threads. This was causing errors with consumer group usage, and was fixed by simply not managing the FileSystem lifecycle (which seems to be the recommended way to handle the FS).